File: src/main/java/com/klarna/rest/api/ApiException.java
package com.klarna.rest.api;

import com.klarna.rest.model.ErrorMessage;
import org.apache.commons.lang3.StringUtils;

/**
 * Generic API exception with the underlying HTTP status and messages.
 */
public class ApiException extends RuntimeException {

    /**
     * The returned HTTP status code for this error.
     */
    private final int httpStatus;

    /**
     * An error message received through the Klarna API.
     *
     * In the event of an error the message contains a
     * {@link com.klarna.rest.model.ErrorMessage#getCorrelationId()}
     * identifying this particular transaction in Klarna's systems.
     *
     * The correlation id may be requested by merchant support to
     * facilitate support inquiries.
     */
    private final ErrorMessage errorMessage;

    /**
     * Constructs a ApiException instance.
     *
     * @param status HTTP status code
     * @param errorMessage Error message
     */
    public ApiException(final int status, final ErrorMessage errorMessage) {
        super(formatError(errorMessage, status));

        this.httpStatus = status;
        this.errorMessage = errorMessage;
    }

    /**
     * Gets the error message.
     *
     * @return Error message
     */
    public ErrorMessage getErrorMessage() {
        return this.errorMessage;
    }

    /**
     * Gets the HTTP status code.
     *
     * @return Status code
     */
    public int getHttpStatus() {
        return this.httpStatus;
    }

    /**
     * Formats the exception message.
     *
     * @param errorMessage Error data
     * @param httpStatus HTTP status code
     * @return Exception message
     */
    private static String formatError(final ErrorMessage errorMessage,
                                      final int httpStatus
    ) {
        return String.format(
                "%s: %s (#%s)",
                errorMessage.getErrorCode(),
                StringUtils.join(errorMessage.getErrorMessages(), ", "),
                errorMessage.getCorrelationId());
    }
}

File: Klarna.Rest/Klarna.Rest.Core/Communication/ApiException.cs
using System;
using System.Net;
using Klarna.Rest.Core.Model;

namespace Klarna.Rest.Core.Communication
{
    /// <summary>
    /// Error message model.
    /// </summary>
    public class ApiException : WebException
    {
        /// <summary>
        /// Gets the API response status code.
        /// </summary>
        public HttpStatusCode StatusCode { get; internal set; }

        /// <summary>
        /// Gets the API error response model.
        /// </summary>
        public ErrorMessage ErrorMessage { get; internal set; }

        /// <summary>
        /// Initializes a new instance of the <see cref="ApiException" /> class.
        /// </summary>
        /// <param name="message">The text of the error message.</param>
        /// <param name="statusCode">The HTTP status code.</param>
        /// <param name="errorMessage">The API error message model.</param>
        /// <param name="innerException">The nested exception</param>
        public ApiException(string message, HttpStatusCode statusCode, ErrorMessage errorMessage, Exception innerException) : base(message, innerException)
        {
            StatusCode = statusCode;
            ErrorMessage = errorMessage;
        }
    }
}
